runtime/src/main/cpp/Cleaner.cpp

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright 2010-2020 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license
+ * that can be found in the LICENSE file.
+ */
+
+#include "Cleaner.h"
+
+#include "Memory.h"
+#include "Runtime.h"
+#include "Worker.h"
+
+// Defined in Cleaner.kt
+extern "C" void Kotlin_CleanerImpl_shutdownCleanerWorker(KInt, bool);
+extern "C" KInt Kotlin_CleanerImpl_createCleanerWorker();
+
+namespace {
+
+struct CleanerImpl {
+    ObjHeader header;
+    KNativePtr cleanerStablePtr;
+};
+
+constexpr KInt kCleanerWorkerUninitialized = 0;
+constexpr KInt kCleanerWorkerInitializing = -1;
+constexpr KInt kCleanerWorkerShutdown = -2;
+
+KInt globalCleanerWorker = kCleanerWorkerUninitialized;
+
+void disposeCleaner(CleanerImpl* thiz) {
+    auto worker = atomicGet(&globalCleanerWorker);
+    RuntimeAssert(worker != kCleanerWorkerUninitialized, "Cleaner worker must've been initialized by now");
+    if (worker == kCleanerWorkerInitializing) {
```

Comment:
Can this really be `true` here? When creating a cleaner, runtime waits until the worker gets initialized.

Reply:
True. Fixed.